src/context.h

```diff
@@ -3,32 +3,70 @@
 #include <map>
 #include <string>
 #include <vector>
+#include <memory>
 #include <unordered_map>
 #include "value.h"
 #include "Assignment.h"
 #include "memory.h"
 
-class Context
+/**
+ * Local handle to a all context objects. This is used to maintain the
+ * dynamic scoping stack using object lifetime.
+ * The Context objects can hang around for longer, e.g. in case of closures.
+ */
+template<typename T>
+struct ContextHandle
 {
+    ContextHandle(std::shared_ptr<T>&& sp) : ctx(std::forward<std::shared_ptr<T>>(sp)) {
+        ctx->init();
+		ctx->push(ctx);
+    }
+
+    ~ContextHandle() {
+        ctx->pop();
+    }
+
+	const T* operator->() const { return ctx.get(); }
+    T* operator->() { return ctx.get(); }
+
+    const std::shared_ptr<T> ctx;
+};
+
+class EvalContext;
+
+class Context : public std::enable_shared_from_this<Context>
+{
+protected:
+	Context(const std::shared_ptr<Context> parent = std::shared_ptr<Context>());
+
 public:
-	typedef std::vector<const Context*> Stack;
+	typedef std::vector<std::shared_ptr<Context>> Stack;
+
+	std::shared_ptr<Context> get_shared_ptr() { return shared_from_this(); }
+	void push(std::shared_ptr<Context> ctx);
+	void pop();
+
+    template<typename C, typename ... T>
+    static ContextHandle<C> create(T&& ... t) {
+        return ContextHandle<C>{std::shared_ptr<C>(new C(std::forward<T>(t)...))};
```

Comment:
Usually, you do `make_shared<C>()` instead of calling the constructor in these cases, which would allow for move semantics if everything else works out.

Reply:
I wanted to prevent construction by other means to ensure the scope stack is maintained via the `ContextHandle` object. With protected constructor or destructor, the `make_shared<C>()` gives an error. Any tips how to get both?

Reply:
Oh, didn't think about the protected constructor. I wouldn't dig too  deep into that - this is where C++  gets really obscure..

Reply:
I found some example of how it would be possible and those look a bit crazy :). After some minimal performance testing, I don't see any change with `make_shared()` (which is supposed to potentially optimize for one allocation instead of two?). So for now I'd prefer the protected constructor to make sure we don't accidentally break the dynamic scoping hierarchy.